pkg/loader/loader.go

```diff
@@ -341,13 +343,87 @@ func LoadRootsWithConfig(cfg *packages.Config, roots ...string) ([]*Package, err
 	// put our build flags first so that callers can override them
 	l.cfg.BuildFlags = append([]string{"-tags", "ignore_autogenerated"}, l.cfg.BuildFlags...)
 
-	rawPkgs, err := packages.Load(l.cfg, roots...)
-	if err != nil {
+	// check each root to see if it should be expanded to include nested modules
+	var goModDirs []string
+	findGoModules := func(p string, d os.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if !d.IsDir() && path.Base(p) == "go.mod" {
+			absPathDirOfGoMod, err := filepath.Abs(filepath.Dir(p))
+			if err != nil {
+				return err
+			}
+			goModDirs = append(goModDirs, absPathDirOfGoMod)
+		}
+		return nil
+	}
+	for i, r := range roots {
```

Comment:
What is the reason for the new four dot notation? Why would someone ever want to use this recursively but not for submodules, if any?

Reply:
The reasoning is to preserve the existing behavior by default. The four-dots enforces an explicit choice to recurse across the boundary.

Reply:
Well sure, but we can break compatibility and if it improves things we should. So then the question becomes which of the two (including submodules or not) do we consider to be more common? That should be the default.

I think it makes a lot of sense to put api packages into a submodule, so I would assume that to be 80% case and not the other way round. So if anything we should probably add a flag to opt out of this (but it might be ok to wait with that until someone actually complains, because until we don't know if anyone even needs this)

Reply:
So one reason I opted to not make this a flag on the CLI was because I was perhaps a little too fond of the "adding an extra dot." I'm not opposed to the opt-out approach, but it _would_ mean that we cannot just start with the four-dot notation and move to opt-out later. Because then we'd break the people that used four-dots (if it was removed).

I suppose I'd like someone else to break the "tie" as it were, but as I've seen mentioned, this is a project whose  consumption far outweighs its developer engagement.

Reply:
I guess I'm in this odd predicament of fighting against my own change 😃 I feel somewhat strongly we should not break things, but if you are okay with it, I am too. My one other concern would be what documentation we need to update and where?

Reply:
Do we have examples of where things would break?  If so... preserve.  Otherwise, change the default and add release notes on the change.

Reply:
I do not, no. Only if people were using Go modules to "hide" certain APIs from this tool. I agree with @alvaroaleman that 80% of people would be fine, and in fact I've been one of the loudest that API dirs _should_ be distinct modules. I just like to err on the side of "don't go breakin' stuff when you don't have to." Then again, I guess I forgot the second part of "move fast..."

Reply:
IMHO it should also be mentioned that if someone for some reason has a submodule within an api directoty that they do indeed not want to include, they can get that by explicitly including all other dirs in their api dir except the one with the submodule - Yes, that takes a non-zero amount of work but it also makes it way more obvious that that submodule shouldn't be included than abusing a bug to get that behavior